# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' PollySI() outputs a matrix of SNP and indel loci at which both the major and minor allele frequency are greater than or equal to an adjustable threshold.
#' @param freq_name A string indicating the VCFtools .frq file. Be sure to include the ".frq" suffix. See Supplemental Materials for instructions on creation.
#' @param desired_scaffolds A vector of scaffolds. Only SNPs and indels on the scaffolds in this vector will be included in the final output.
#' @param threshold Numeric. The minimum frequency an allele must have to be considered highly polymorphic.
PollySI <- function(freq_name, desired_scaffolds, threshold) {
    .Call(`_Polly_PollySI`, freq_name, desired_scaffolds, threshold)
}

#' MicroGenotyper() finds the genotypes of each specified individual at each of the microsatellite loci indicated by the lookup table.
#' @param bam_vec A vector of .bam files, each from a separate individual, to be genotyped.
#' @param lookup_file A string denoting the .csv in which the known microsatellite loci for the species are stored. See Supplemental Materials for instructions on creation.
#' @param desired_scaffolds A vector of scaffolds. Only microsatellites on these scaffolds will be genotyped.
#' @param output_names A vector of CSV files into which the microsatellite genotypes of each .bam file will be outputted. Must be in the same order as the input names specified in bam_vec. Be sure to include ".csv" at the end of each name.
MicroGenotyper <- function(bam_vec, lookup_file, desired_scaffolds, output_names) {
    invisible(.Call(`_Polly_MicroGenotyper`, bam_vec, lookup_file, desired_scaffolds, output_names))
}

#' Identifies user-defined polymorphic microsatellites from the output of MicroGenotyper.
#' @param CSV_names A vector of CSV output files from MicroGenotyper. Be sured to include ".csv" to the end of each name.
#' @param desired_scaffolds A vector of scaffolds. Only microsatellites on these scaffolds will be checked for polymorphism.
#' @param threshold An integer describing the minimum number of alleles a microsatellite must have in the population to be considered polymorphic.
#' @param output_name A string naming the CSV file into which polymorphic microsatellites ought to be listed. Be sure to include ".csv" suffix.
PollyMicros <- function(CSV_names, desired_scaffolds, threshold, output_name) {
    invisible(.Call(`_Polly_PollyMicros`, CSV_names, desired_scaffolds, threshold, output_name))
}

#' Combines the polymorphic microsatellites detected by PollyMicros() with the the polymorphic SNPs and indels detected by PollySI() to group polymorphic microsatellites, SNPs, and indels into regions of a user-defined size.
#' @param region_size Size (in base pairs) of regions into which polymorphic microsatellites, SNPs, and indels ought to be grouped.
#' @param desired_scaffolds A vector of scaffolds. Only microsatellites, SNPs, and indels on these scaffolds will be genotyped and checked for polymorphism.
#' @param PollySIInputName A string indicating the name of the CSV output of PollySI(). Be sure to include the ".csv" suffix.
#' @param PollyMicrosInputName A string indicating the name of the CSV output of PollyMicros(). Be sure to include the ".csv" suffix.
#' @param FinalOutputName A string naming the CSV file into which the highly-polymorphic SNP, indel, and microsatellite clusters ought to be outputted. Be sure to include the ".csv" suffix.
Polly <- function(region_size, desired_scaffolds, PollySIInputName, PollyMicrosInputName, FinalOutputName) {
    invisible(.Call(`_Polly_Polly`, region_size, desired_scaffolds, PollySIInputName, PollyMicrosInputName, FinalOutputName))
}

